src/test/unittest/ut_signal.c

```diff
@@ -40,12 +40,31 @@
 /*
  * On Windows, Access Violation exception does not raise SIGSEGV signal.
  * The trick is to catch the exception and... call the signal handler.
- *
- * XXX - add support for registering more than one signal/exception handler
  */
 
-static void (*Sa_handler) (int signum);
-static int Signum;
+struct signal_handler {
```

Comment:
This is needlessly complicated. And incorrect.

There are 32 signals that we are interested in, which translates to a static array of 32 sigaction structs and a single static lock;

sigaction:
lock(&action_lock);
if (oldact)
  *oldact = actions[signum - 1];
if (act)
  actions[signum - 1] = *act;
unlock(&action_lock);
return 0;

in the handler code:
actions[signum - 1].sa_handler(signum);

done.

To properly implement sigaction you should also take mask into account make this thing per-thread, but I'm not sure if we want to go that far.